### PR TITLE
1 - remove the bundled binaries, 2 - download binaries on startup

### PR DESCRIPTION
**Pull request by dimacodota**
1 - remove the bundled binaries, 2 - download binaries on startup(check vscode proxy settings)

### CODE REVIEW

File: src/binary/binaryFetcher/bundleDownloader.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  await createBundleDirectory(bundleDirectory);
+  await downloadFileToDestination(bundleDownloadUrl, bundlePath);
+  await extractBundle(bundlePath, bundleDirectory);
+  await setDirectoryFilesAsExecutable(bundleDirectory);
+  return executablePath;
+}
+
+async function getBundlePaths(): Promise<BundlePaths> {
+  const version = await getCurrentVersion();
+  const bundlePath = geBundlePath(version);
```
Comment (bilucodota) [inline src/binary/binaryFetcher/bundleDownloader.ts:40]: typo in `geBundlePath`

File: src/binary/binaryFetcher/bundleDownloader.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  return fs.unlink(bundle);
+}
+
+async function setDirectoryFilesAsExecutable(
+  bundleDirectory: string
+): Promise<void[]> {
+  const files = await fs.readdir(bundleDirectory);
+  return Promise.all(
+    files.map((file) =>
+      fs.chmod(path.join(bundleDirectory, file), EXECUTABLE_FLAG)
```
Comment (bilucodota) [inline src/binary/binaryFetcher/bundleDownloader.ts:70]: Is it working on `Windows`?

File: src/proxySupport.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    return { agent: undefined, rejectUnauthorized: false };
+  }
+
+  const rejectUnauthorized = workspace
+    .getConfiguration()
+    .get("http.proxyStrictSSL", true);
+
+  const proxyOptions = {
+    host: proxyUrl.hostname,
+    port: parseInt(proxyUrl.port || "", 10),
```
Comment (bilucodota) [inline src/proxySupport.ts:27]: `parseInt("")` is dangerous here no?

Comment (dimacodota) [reply src/proxySupport.ts]: oh... obviously, good catch

File: src/binary/binaryFetcher/bundleDownloader.ts (hunk 2 above)
Comment (dimacodota) [reply src/binary/binaryFetcher/bundleDownloader.ts:70]: yep

File: package.json [hunk 4]
```diff
@@ -115,11 +115,14 @@
...
     "@types/glob": "^7.1.3",
     "@types/mocha": "^8.2.0",
+    "@types/mock-fs": "^4.13.0",
+    "@types/ncp": "^2.0.4",
     "@types/node": "^10.0.0",
     "@types/semver": "^7.3.4",
     "@types/sinon": "^9.0.10",
     "@types/tmp": "^0.2.0",
     "@types/vscode": "^1.29.0",
+    "@types/yauzl": "^2.9.1",
```
Comment (boaz-codota) [inline package.json:125]: you installed the `@types/yazul` but where is the library itself? Why isn't it a dependency as well?

File: package.json [hunk 5]
```diff
@@ -146,6 +150,8 @@
   },
   "dependencies": {
     "await-semaphore": "^0.1.3",
+    "extract-zip": "^2.0.1",
```
Comment (boaz-codota) [inline package.json:153]: Which one are you using? this one or `yauzl`?

File: src/binary/binaryFetcher/binaryValidator.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import * as fs from "fs";
+import { runProcess } from "../runProcess";
+// import { runProcess } from "../runProcess";
```
Comment (boaz-codota) [inline src/binary/binaryFetcher/binaryValidator.ts:3]: remove

File: src/binary/binaryFetcher/binaryValidator.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        resolve(false);
+      }
+    });
+
+    proc.on("error", () => {
+      resolve(false);
+    });
+
+    readLine.once("line", (line: string) => {
+      if (line) {
```
Comment (boaz-codota) [inline src/binary/binaryFetcher/binaryValidator.ts:29]: why not read the line?

File: src/binary/binaryFetcher/binaryValidator.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  const exists = fs.existsSync(version);
+  if (!exists) {
+    return false;
+  }
+  const { proc, readLine } = runProcess(version, ["--print-version"]);
+
+  return new Promise((resolve) => {
+    setTimeout(() => {
+      console.error(`validating ${version} timeout out`);
+      resolve(false);
```
Comment (boaz-codota) [inline src/binary/binaryFetcher/binaryValidator.ts:15]: why use resolve with boolean over using reject?

File: src/binary/binaryFetcher/binaryValidator.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
+import * as fs from "fs";
+import { runProcess } from "../runProcess";
+// import { runProcess } from "../runProcess";
+
+export default async function isValidBinary(version: string): Promise<boolean> {
+  const exists = fs.existsSync(version);
```
Comment (boaz-codota) [inline src/binary/binaryFetcher/binaryValidator.ts:6]: why sync?

File: src/binary/binaryFetcher/existingVersionHandler.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+export default async function handleExistingVersion(
+  rootPath: string
+): Promise<string | null> {
+  try {
+    if (fs.existsSync(rootPath)) {
+      const versions = sortBySemver(fs.readdirSync(rootPath)).map(versionPath);
+      const selectedVersion = await asyncFind(versions, isValidBinary);
+      if (selectedVersion) {
+        return selectedVersion;
+      }
```
Comment (boaz-codota) [inline src/binary/binaryFetcher/existingVersionHandler.ts:16]: just return the result of asyncFind, the rest is redundant

File: src/utils.ts [hunk 11]
```diff
@@ -41,3 +41,18 @@ export function isFunction(functionToCheck: any): boolean {
...
+  return arr.reduce(async (result: Promise<T | null>, element: T) => {
+    if (await result) {
+      return result;
+    }
+    if (await predicate(element)) {
+      return element;
+    }
+    return Promise.resolve(null);
+  }, Promise.resolve(null));
+}
```
Comment (boaz-codota) [inline src/utils.ts:58]: I can't understand what this function does, and whether or not it does it as intended. Can you refactor it to not use `reduce`?

File: src/test/suite/utils/fs.mock.ts [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import * as fs from "fs";
+import * as sinon from "sinon";
+
+export default function mockExistsSync(args: [fs.PathLike, boolean][]): void {
```
Comment (boaz-codota) [inline src/test/suite/utils/fs.mock.ts:4]: You can use the `...args` syntax to expose a cleaner api: see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments#rest_default_and_destructured_parameters

File: src/test/suite/fetchBinaryPath.test.ts [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+    expect(version).to.equal(expectedVersion);
+  });
+  test("if no .active and no version exists, return the downloaded version", async () => {
+    const versionToDownload = "1.2.5";
+
+    mockExistsSync([
+      [versionPath(versionToDownload), true],
+      [getRootPath(), false],
+    ]);
```
Comment (boaz-codota) [inline src/test/suite/fetchBinaryPath.test.ts:53]: the `true` and `false` here are not readable. Could you replace the `boolean` param with an `enum` with meaningful names?

File: src/test/suite/fetchBinaryPath.test.ts [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+      [versionPath(currentVersion)]: "test binary",
+    });
+
+    const version = await fetchBinaryPath();
+    const expectedVersion = versionPath(currentVersion);
+
+    expect(version).to.equal(expectedVersion);
+  });
+  test("if no .active and no version exists, return the downloaded version", async () => {
+    const versionToDownload = "1.2.5";
```
Comment (boaz-codota) [inline src/test/suite/fetchBinaryPath.test.ts:48]: move this const (and the others) to another file (testData etc.). It is implementation detail that makes it harder to read the test

Comment (boaz-codota) [inline src/test/suite/fetchBinaryPath.test.ts:48]: move this const (and the others) to another file (testData etc.). It is implementation detail that makes it harder to read the test

File: src/test/suite/fetchBinaryPath.test.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+      [versionPath(versionToDownload), true],
+      [getRootPath(), false],
+    ]);
+
+    mockHttp([
+      [versionToDownload, getUpdateVersion()],
+      [
+        fs.createReadStream(
+          path.join(__dirname, "..", "fixture", "TabNine.zip")
+        ),
```
Comment (boaz-codota) [inline src/test/suite/fetchBinaryPath.test.ts:60]: extract method

File: src/test/runTest.ts [hunk 16]
```diff
@@ -16,6 +17,16 @@ async function main() {
     // Passed to --extensionTestsPath
     const extensionTestsPath = path.resolve(__dirname, "./suite/index");
 
+    const fixtureBinary = path.resolve(__dirname, "./fixture/binaries");
+    const targetBinary = path.resolve(__dirname, "../..", "binaries");
+
+    ncp(fixtureBinary, targetBinary, (err) => {
```
Comment (boaz-codota) [inline src/test/runTest.ts:23]: what about cleaning after the tests all finished running?

**Review by boaz-codota — COMMENTED** [review]
Had some feedback

**Review by boaz-codota — APPROVED** [review]

**Comment by boaz-codota** [comment]
Do notice that the tests fail on ubuntu

File: package.json (hunk 4 above)
Comment (dimacodota) [reply package.json:125]: for some reason this is required in order to use `extract-zip` when compiled on linux

File: package.json (hunk 5 above)
Comment (dimacodota) [reply package.json:153]: extract-zip

File: src/binary/binaryFetcher/binaryValidator.ts (hunk 9 above)
Comment (dimacodota) [reply src/binary/binaryFetcher/binaryValidator.ts:6]: oh, this was the original code, ill change it

File: src/binary/binaryFetcher/binaryValidator.ts (hunk 8 above)
Comment (dimacodota) [reply src/binary/binaryFetcher/binaryValidator.ts:15]: reject will break the execution with exception  until it caught, and here  i want to go over each directory and check if it is valid

File: src/binary/binaryFetcher/binaryValidator.ts (hunk 7 above)
Comment (dimacodota) [reply src/binary/binaryFetcher/binaryValidator.ts:29]: what do you mean ?

**Review by boaz-codota — APPROVED** [review]

**Pull request merged**